### ISSUE TITLE
The connection observed an error

### ISSUE THREAD

**Issue by Rupaligangarde** [reactor/reactor-netty#1621]
Versions:
SpringBootVersion: 2.4.5
io.projectreactor.netty:reactor-netty-core:1.0.6
org.springframework.boot:spring-boot-starter-reactor-netty:2.4.5
io.netty:netty-*: 4.1.63.Final

I'm using Spring webclient to connect to the server staging.services to POST and GET data. There is no timeout configured at this server side.
I see lots of logs with "The connection observed an error" with multiple different reasons

```
1. "readAddress(..) failed: Connection reset by peer" "io.netty.channel.unix.Errors$NativeIoException"
2. "java.nio.channels.ClosedChannelException"
3. "Connection prematurely closed BEFORE response; nested exception is reactor.netty.http.client.
: Connection prematurely closed BEFORE response"
```
None of the request is successful.

Here is my webclient config:
```
ConnectionProvider connectionProvider = ConnectionProvider.builder("catalogPool")
                .maxConnections(1000)
                .maxIdleTime(Duration.ofMinutes(1))
                .pendingAcquireTimeout(Duration.ofMillis(75000))
                .build();

        HttpClient httpClient = HttpClient.create(connectionProvider)
                .doOnConnected(conn -> conn
                        .addHandler(new ReadTimeoutHandler(30, TimeUnit.SECONDS))
                        .addHandler(new WriteTimeoutHandler(30))
                );

        this.webClient = webClientBuilder
                .baseUrl(baseUrl)
                .build();
```

tried with preconfigured Webclient.Builder way as well. Results in Same error.
POST call details:

 ```
public <T, R> R postData(T data, String tenant, String uri, Class<R> clazz) {
        Map<String, String> headers = headers(tenant);
        return webClient
                .post()
                .uri(uri)
                .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
                .header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE)
                .headers(httpHeaders -> headers.forEach(httpHeaders::set))
                .body(BodyInserters.fromValue(data))
                .retrieve()
                .onStatus(s -> s != HttpStatus.OK, response -> {
                    Mono<WebClientResponseException> exception = response.createException();
                    response.releaseBody();
                    return exception;
                })
                .toEntity(clazz)
                .map(HttpEntity::getBody);
    }
```
GET call details:

 ```
public <R> List<R> getAllData(String tenant, String uri, Class<R> clazz) {
        Map<String, String> headers = headers(tenant);
        return webClient
                .get()
                .uri(uri)
                .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_VALUE)
                .header(HttpHeaders.ACCEPT, MediaType.APPLICATION_JSON_VALUE)
                .headers(httpHeaders -> headers.forEach(httpHeaders::set))
                .retrieve()
                .onStatus(s -> s != HttpStatus.OK, response -> {
                    Mono<WebClientResponseException> exception = response.createException();
                    response.releaseBody();
                    return exception;
                })
                .toEntityList(clazz)
                .map(HttpEntity::getBody);
    }
```

I have detailed debug logs here. I tried all all answers and config mentioned in official documents. but error still persists, really need help/inputs on this error.

Have attached detailed logs. Please revert.
[downloaded-logs-20210503-235837 (2).csv](https://github.com/reactor/reactor-netty/files/6427736/downloaded-logs-20210503-235837.2.csv)

**Comment by Rupaligangarde** [comment]
Hello all, 

to understand my issue, i have created one sample application with springboot 2.4.3. since connection were not released properly in 2.4.5 version.

I have "/test" endpoint which internally make 1000 parallel GET calls to some remote server using webclient configuration as below:
```
ConnectionProvider connectionProvider = ConnectionProvider.builder("SamplePool")
                .maxConnections(100)
                .maxLifeTime(Duration.ofMinutes(3L))
                .build();

        HttpClient httpClient = HttpClient.create(connectionProvider);
        webClient = WebClient.builder()
                .clientConnector(new ReactorClientHttpConnector(httpClient))
                .baseUrl("https://staging.some.services")
                .build();
```

Local logs are in the attched file, filename: localLogs.txt. It releases connection properly and uses webclient thread pool (threadid starts with ctor-http-nio-*) if i understand it right. Returns 200 response.
When I ran this sample project on env, I see it does not release connection properly ans uses some other threads(epoll) and returns 500 response. logs in envLogs.txt
I believe this is caaused because of some mismatch dependency config, based on this github issue: https://github.com/reactor/reactor-netty/issues/1368

But In my scenario, there is no load, remote server is same, endpoint is same, no of GET calls made are same. dependency versions are same, code is same, I am not sure why is this happening.
Can you please confirm same for me? I have mentioned gradle dependencies o/p in dep.txt file.
[envLogs.txt](https://github.com/reactor/reactor-netty/files/6434489/envLogs.txt)
[localLogs.txt](https://github.com/reactor/reactor-netty/files/6434490/localLogs.txt)
[dep.txt](https://github.com/reactor/reactor-netty/files/6434494/dep.txt)

**Comment by violetagg** [comment]
@Rupaligangarde Let me summarise my understanding:
- the code for the client in both cases is the same
- the endpoint (at least the IPs) is different
- when you run from a local machine to some endpoint the code is working, when you move the same code to some other environment you see connections are dropped

Do you have some network component in that environment that may stop the traffic?
Can you capture the traffic with Wireshark or similar and share it/inspect it in order to seem which party drops the connections?

In the local environment you see  `ctor-http-nio-*` because most probably you do not have native transport and NIO transport is used instead, while in the other environment you use Epoll instead of NIO. However this should not cause problems. In order to check that you can run with `-Dreactor.netty.native=false`.

**Comment by Rupaligangarde** [comment]
Hi Violeta,
Thanks for the reply.

- the code for the client in both cases is the same => Yes, code is same.
- the endpoint (at least the IPs) is different in the log files => Yes, I have changed the actual IPs in all the files to some random numbers for confidentiality. But It is exact same server i am referring in entire discussions/files. 
- when you run from a local machine to some endpoint the code is working, when you move the same code to some other environment you see connections are dropped => Yes, I see connection error observed an error logs.

Do you have some network component in that environment that may stop the traffic?
- there is no component place in between my service(client) and server.

Can you capture the traffic with Wireshark or similar and share it/inspect it in order to see which party drops the connections?
=> sure, I will send.

Thanks for suggesting `-Dreactor.netty.native=false` option, will try it once.

Are dependencies used are correct for 2.4.3 version?

**Comment by Rupaligangarde** [comment]
Hi Violeta,
I can not wireshark req/res and share those logs, but i managed to reproduce below error locally. 
```
2021-05-07 14:50:40.350  WARN 49752 --- [tor-http-nio-11] r.netty.http.client.HttpClientConnect    : [id: 0xad7ae156, L:/134.0.2.45:51542 ! R:staging.some.services/108.11.6.44:443] The connection observed an error
reactor.netty.http.client.PrematureCloseException: Connection prematurely closed BEFORE response
```

I have also capture logs using httpclient.wiretap option. I hope this helps us. Please find logs [here](https://drive.google.com/file/d/14Z975c7JMLCMgMginuf5lYPrV7s9ufHJ/view?usp=sharing)
webclient config used to this is

```
public ExternalService(
            @Autowired WebClient.Builder webClientBuilder
    ) {

        ConnectionProvider connectionProvider = ConnectionProvider.create("SamplePool", 100);

        HttpClient httpClient = HttpClient.create(connectionProvider)
                .wiretap(true)
                .responseTimeout(Duration.ofMinutes(3L));

        webClient = webClientBuilder
                .clientConnector(new ReactorClientHttpConnector(httpClient))
                .baseUrl("https://staging.some.services")
                .build();
    }
```

**Comment by violetagg** [comment]
@Rupaligangarde when you use `response.releaseBody()` you have to subscribe.
```
.onStatus(HttpStatus::isError, response -> {
                    Mono<Throwable> exception = response.createException().flatMap(Mono::error);
                    response.releaseBody();
                    return exception;
                })
```

**Comment by violetagg** [comment]
> I can not wireshark req/res and share those logs

You can try with [tcpdump](https://en.wikipedia.org/wiki/Tcpdump)

**Comment by Rupaligangarde** [comment]
[tcpdump.txt](https://github.com/reactor/reactor-netty/files/6451363/tcpdump.txt)
"2021-05-10 10:24:59.498  WARN 1 --- [ctor-http-nio-3] r.n.h.c.HttpClientConnect                : [id: 0x23ebea30, L:/10.22.1.34:56102 - R:staging.falabella.services/104.17.239.83:443] The connection observed an error
"

**Comment by violetagg** [comment]
@Rupaligangarde Except this exception `ClosedChannelException` do you see some other exception? I'm searching for this port `56102` in the tcpdump but I do not see neither [F] nor [R] flags.

**Comment by Rupaligangarde** [comment]
[Uploading downloaded-logs-20210510-164554.csv…]()

**Comment by violetagg** [comment]
@Rupaligangarde configure the ConnectionProvider like this

```
ConnectionProvider provider = ConnectionProvider.builder("custom")
        .maxConnections(10)
        .pendingAcquireMaxCount(-1)
        .build();
```

**Comment by Rupaligangarde** [comment]
@violetagg Thanks for the reply.. I tried provided config but still same error.

**Comment by violetagg** [comment]
@Rupaligangarde which error? `connection prematurely closed` or `Pending acquire queue has reached its maximum size of 20`

**Comment by violetagg** [comment]
@Rupaligangarde We need an update in order to continue with the investigation.

**Comment by Rupaligangarde** [comment]
@violetagg 
Hello Violeta,
I did some more debugging on same connection error. As a part of debugging, I switched from webclient to rest template approach and i saw `52x`  exceptions. All connection errors i was seeing are 52x exceptions only.
this is happening at proxy level ( cloudFlare).  I think, these errors are nothing to do with reactor netty but actual errors causing at proxy level.

`i think Netty just wraps these exception into conn observed an error exception.. under the hood, these are 521 and 525 exceptions. (UnknownHttpStatusCodeException)`

It would be better if netty/webclient also throws clear httpStatus code & error message similar to rest template `UnknownHttpStatusCodeException` instead of wrapping everything as `conn observed an error exception`
I am not sure if its there as a part of enhancement, but it will be great help if we receive something similar to UnknownHttpStatusCodeException when using webclient as well.
